2.0 - intellijPlatform in settings plugin tries to access project instance

### What happened?

```kotlin
// myRepos.settings.gradle.kts
import org.jetbrains.intellij.platform.gradle.extensions.intellijPlatform

plugins {
    id("org.jetbrains.intellij.platform.settings")
}

dependencyResolutionManagement {
    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
    repositories {
        mavenCentral()
        maven(url = "https://oss.sonatype.org/content/repositories/snapshots")
        maven(url = "https://s01.oss.sonatype.org/content/repositories/snapshots")

        intellijPlatform {
            defaultRepositories()
        }
    }
}
```

### Relevant log output or stack trace

```shell
Caused by: java.lang.IllegalStateException: The root project is not yet available for build.
	at org.gradle.invocation.DefaultGradle.getRootProject(DefaultGradle.java:258)
	at org.gradle.invocation.DefaultGradle_Decorated.getRootProject(Unknown Source)
	at org.gradle.invocation.DefaultGradle.getRootProject(DefaultGradle.java:82)
	at org.jetbrains.intellij.platform.gradle.extensions.ExtensionsKt.getProjectCacheDirectory(extensions.kt:84)
	at org.jetbrains.intellij.platform.gradle.extensions.ExtensionsKt.getLocalPlatformArtifacts(extensions.kt:98)
	at org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesExtension$localPlatformArtifacts$2.execute(IntelliJPlatformRepositoriesExtension.kt:182)
	at org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesExtension$localPlatformArtifacts$2.execute(IntelliJPlatformRepositoriesExtension.kt:180)
	at org.gradle.api.internal.artifacts.DefaultArtifactRepositoryContainer.addRepository(DefaultArtifactRepositoryContainer.java:88)
	at org.gradle.api.internal.artifacts.dsl.DefaultRepositoryHandler.ivy(DefaultRepositoryHandler.java:172)
	at org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesExtension.localPlatformArtifacts(IntelliJPlatformRepositoriesExtension.kt:180)
	at org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesExtension.localPlatformArtifacts$default(IntelliJPlatformRepositoriesExtension.kt:180)
	at org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesExtension.defaultRepositories(IntelliJPlatformRepositoriesExtension.kt:216)
	at MyRepos_settings_gradle$2$1$1.execute(myRepos.settings.gradle.kts:15)
```


### Steps to reproduce

1. Check out https://github.com/hfhbd/kobol/tree/intellij-2.0.0
2. `./gradlew :help`

### Gradle IntelliJ Plugin version

2.0.0

### Gradle version

8.6

### Operating System

Linux

### Link to build, i.e. failing GitHub Action job

_No response_

--- a/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformDependenciesExtension.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformDependenciesExtension.kt
@@ -7,7 +7,7 @@ import org.gradle.api.artifacts.Dependency
 import org.gradle.api.artifacts.dsl.DependencyHandler
 import org.gradle.api.artifacts.dsl.RepositoryHandler
 import org.gradle.api.file.Directory
-import org.gradle.api.invocation.Gradle
+import org.gradle.api.file.ProjectLayout
 import org.gradle.api.provider.Provider
 import org.gradle.api.provider.ProviderFactory
 import org.gradle.internal.os.OperatingSystem
@@ -49,7 +49,7 @@ import kotlin.math.absoluteValue
  * @param repositories The Gradle [RepositoryHandler] to manage repositories.
  * @param dependencies The Gradle [DependencyHandler] to manage dependencies.
  * @param providers The Gradle [ProviderFactory] to create providers.
- * @param gradle The [Gradle] instance.
+ * @param rootProjectDirectory The [Directory] of the root project.
  */
 @Suppress("unused", "MemberVisibilityCanBePrivate")
 @IntelliJPlatform
@@ -58,7 +58,8 @@ abstract class IntelliJPlatformDependenciesExtension @Inject constructor(
     private val repositories: RepositoryHandler,
     private val dependencies: DependencyHandler,
     private val providers: ProviderFactory,
-    private val gradle: Gradle,
+    private val layout: ProjectLayout,
+    private val rootProjectDirectory: File,
 ) {
 
     /**
@@ -748,14 +749,17 @@ abstract class IntelliJPlatformDependenciesExtension @Inject constructor(
             productInfo.validateSupportedVersion()
 
             val type = productInfo.productCode.toIntelliJPlatformType()
-            val hash = artifactPath.absolutePathString().hashCode().absoluteValue % 1000
+            val hash = artifactPath.hashCode().absoluteValue % 1000
 
             dependencies.create(
                 group = Configurations.Dependencies.LOCAL_IDE_GROUP,
                 name = type.dependency.name,
                 version = "${productInfo.version}+$hash",
             ).apply {
-                createIvyDependency(gradle, listOf(artifactPath.toPublication()))
+                createIvyDependency(
+                    providers.localPlatformArtifactsDirectory(rootProjectDirectory),
+                    listOf(artifactPath.toPublication())
+                )
             }.apply(action)
         },
     )
@@ -857,7 +861,10 @@ abstract class IntelliJPlatformDependenciesExtension @Inject constructor(
             name = id,
             version = "${productInfo.version}+$hash",
         ).apply {
-            createIvyDependency(gradle, jars.map { it.toPublication() })
+            createIvyDependency(
+                providers.localPlatformArtifactsDirectory(rootProjectDirectory),
+                jars.map { it.toPublication() }
+            )
         }
     }
 

--- a/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformExtension.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformExtension.kt
@@ -588,10 +588,11 @@ interface IntelliJPlatformExtension : ExtensionAware {
             internal val dependencies: DependencyHandler,
             internal val downloadDirectory: DirectoryProperty,
             internal val extensionProvider: Provider<IntelliJPlatformExtension>,
-            internal val gradle: Gradle,
             internal val providers: ProviderFactory,
+            project: Project,
             internal val resources: ResourceHandler,
         ) {
+            private val rootDir = project.rootDir
 
             /**
              * Adds a dependency to a binary IDE release to be used for testing with the IntelliJ Plugin Verifier.
@@ -664,7 +665,10 @@ interface IntelliJPlatformExtension : ExtensionAware {
              *
              * @param localPath The provider for the type of the IntelliJ Platform dependency. Accepts either [String], [File], or [Directory].
              */
-            fun localIde(localPath: Provider<*>) = addLocalIdeDependency(localPath)
+            fun localIde(localPath: Provider<*>) = addLocalIdeDependency(
+                localPlatformArtifactsDirectory = providers.localPlatformArtifactsDirectory(rootDir),
+                localPath = localPath,
+            )
 
             /**
              * Adds the local IDE to be used for testing with the IntelliJ Plugin Verifier.
@@ -749,7 +753,10 @@ interface IntelliJPlatformExtension : ExtensionAware {
             /**
              * Creates and adds a local instance of the IDE as a dependency.
              */
-            private fun addLocalIdeDependency(localPath: Provider<*>) =
+            private fun addLocalIdeDependency(
+                localPlatformArtifactsDirectory: File,
+                localPath: Provider<*>,
+            ) =
                 configurations.getByName(Configurations.INTELLIJ_PLUGIN_VERIFIER_IDES_LOCAL_INSTANCE).dependencies.addLater(localPath.map {
                     val artifactPath = resolveArtifactPath(it)
                     val productInfo = artifactPath.productInfo()
@@ -764,7 +771,7 @@ interface IntelliJPlatformExtension : ExtensionAware {
                         name = type.dependency.name,
                         version = "${productInfo.version}+$hash",
                     ).apply {
-                        createIvyDependency(gradle, listOf(artifactPath.toPublication()))
+                        createIvyDependency(localPlatformArtifactsDirectory, listOf(artifactPath.toPublication()))
                     }
                 })
         }

--- a/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformRepositoriesExtension.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformRepositoriesExtension.kt
@@ -3,24 +3,19 @@
 package org.jetbrains.intellij.platform.gradle.extensions
 
 import org.gradle.api.Action
-import org.gradle.api.Project
 import org.gradle.api.artifacts.dsl.RepositoryHandler
 import org.gradle.api.artifacts.repositories.ArtifactRepository
-import org.gradle.api.initialization.resolve.DependencyResolutionManagement
-import org.gradle.api.invocation.Gradle
 import org.gradle.api.plugins.ExtensionAware
 import org.gradle.api.provider.ProviderFactory
-import org.gradle.internal.os.OperatingSystem
 import org.gradle.kotlin.dsl.maven
 import org.jetbrains.intellij.platform.gradle.BuildFeature
 import org.jetbrains.intellij.platform.gradle.Constants.Configurations
 import org.jetbrains.intellij.platform.gradle.Constants.Extensions
 import org.jetbrains.intellij.platform.gradle.Constants.Locations
 import org.jetbrains.intellij.platform.gradle.IntelliJPlatformType
-import org.jetbrains.intellij.platform.gradle.model.toPublication
+import java.io.File
 import java.net.URI
 import javax.inject.Inject
-import kotlin.io.path.absolutePathString
 
 /**
  * This is an extension class for managing IntelliJ Platform repositories in a Gradle build script. It's applied to the [RepositoryHandler].
@@ -42,7 +37,7 @@ import kotlin.io.path.absolutePathString
 abstract class IntelliJPlatformRepositoriesExtension @Inject constructor(
     private val repositories: RepositoryHandler,
     private val providers: ProviderFactory,
-    private val gradle: Gradle,
+    private val localPlatformArtifactsDirectory: File,
 ) {
 
     /**
@@ -179,20 +174,11 @@ abstract class IntelliJPlatformRepositoriesExtension @Inject constructor(
      */
     fun localPlatformArtifacts(action: RepositoryAction = {}) = repositories.ivy {
         // Location of Ivy files generated for the current project.
-        ivyPattern("${gradle.localPlatformArtifacts.absolutePathString()}/[organization]-[module]-[revision].[ext]")
+        setUrl(localPlatformArtifactsDirectory.toPath().toUri())
+        ivyPattern("/[organization]-[module]-[revision].[ext]")
 
         // As all artifacts defined in Ivy repositories have a full artifact path set as their names, we can use them to locate artifact files
         artifactPattern("/[artifact]")
-
-        /**
-         * Because artifact paths always start with `/` (see [toPublication] for details),
-         * on Windows, we have to guess to which drive letter the artifact path belongs to.
-         * To do so, we add all drive letters (`a:/[artifact]`, `b:/[artifact]`, `c:/[artifact]`, ...) to the stack,
-         * starting with `c` for the sake of micro-optimization.
-         */
-        if (OperatingSystem.current().isWindows) {
-            (('c'..'z') + 'a' + 'b').forEach { artifactPattern("$it:/[artifact]") }
-        }
     }.apply {
         repositories.exclusiveContent {
             forRepositories(this@apply)

--- a/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/extensions.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/extensions.kt
@@ -4,7 +4,7 @@ package org.jetbrains.intellij.platform.gradle.extensions
 
 import org.gradle.api.artifacts.ExternalModuleDependency
 import org.gradle.api.file.Directory
-import org.gradle.api.invocation.Gradle
+import org.gradle.api.provider.ProviderFactory
 import org.gradle.internal.os.OperatingSystem
 import org.jetbrains.intellij.platform.gradle.BuildException
 import org.jetbrains.intellij.platform.gradle.Constants.GradleProperties
@@ -29,9 +29,12 @@ annotation class IntelliJPlatform
  * @param gradle The [Gradle] instance.
  * @param publications The list of [IvyModule.Publication] objects to be included in the Ivy file.
  */
-internal fun ExternalModuleDependency.createIvyDependency(gradle: Gradle, publications: List<IvyModule.Publication>) {
-    val ivyFile = gradle.localPlatformArtifacts
-        .resolve("$group-$name-$version.xml")
+internal fun ExternalModuleDependency.createIvyDependency(
+    localPlatformArtifactsDirectory: File,
+    publications: List<IvyModule.Publication>,
+) {
+    val ivyFile = localPlatformArtifactsDirectory.toPath()
+        .div("$group/$name/$version/ivy-$version.xml")
         .takeUnless { it.exists() }
         ?: return
 
@@ -78,10 +81,9 @@ internal fun resolveArtifactPath(localPath: Any) = when (localPath) {
     .throwIfNull { BuildException("Specified localPath '$localPath' doesn't exist or is not a directory") }
 
 /**
- * Returns the Gradle project cache directory.
+ * Represents the local platform artifacts directory path which contains Ivy XML files.
  */
-private val Gradle.projectCacheDirectory
-    get() = (startParameter.projectCacheDir ?: this.rootProject.projectDir).toPath().resolve(".gradle")
+internal const val localPlatformArtifactsDirectory = ".gradle/intellijPlatform/ivy"
 
 /**
  * Represents the local platform artifacts directory path which contains Ivy XML files.
@@ -90,9 +92,14 @@ private val Gradle.projectCacheDirectory
  * @see [IntelliJPlatformRepositoriesExtension.localPlatformArtifacts]
  * @see [GradleProperties.LOCAL_PLATFORM_ARTIFACTS]
  */
-internal val Gradle.localPlatformArtifacts
-    get() = runCatching { rootProject.findProperty(GradleProperties.LOCAL_PLATFORM_ARTIFACTS)?.toString() }
-        .getOrNull()
-        .takeUnless { it.isNullOrEmpty() }
-        ?.let { Path(it) }
-        ?: projectCacheDirectory.resolve("intellijPlatform/ivy")
+internal fun ProviderFactory.localPlatformArtifactsDirectory(
+    rootProjectDir: File,
+): File {
+    // gradleProperties are always safe to call get/orNull
+    val customLocalPlatformArtifactsDirectory = gradleProperty(GradleProperties.LOCAL_PLATFORM_ARTIFACTS).orNull
+    return if (customLocalPlatformArtifactsDirectory != null) {
+        File(customLocalPlatformArtifactsDirectory)
+    } else {
+        rootProjectDir.resolve(localPlatformArtifactsDirectory)
+    }
+}

--- a/src/main/kotlin/org/jetbrains/intellij/platform/gradle/plugins/project/IntelliJPlatformBasePlugin.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/platform/gradle/plugins/project/IntelliJPlatformBasePlugin.kt
@@ -27,6 +27,7 @@ import org.jetbrains.intellij.platform.gradle.artifacts.transform.applyPluginVer
 import org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformDependenciesExtension
 import org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformExtension
 import org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesExtension
+import org.jetbrains.intellij.platform.gradle.extensions.localPlatformArtifactsDirectory
 import org.jetbrains.intellij.platform.gradle.isBuildFeatureEnabled
 import org.jetbrains.intellij.platform.gradle.model.platformPath
 import org.jetbrains.intellij.platform.gradle.model.productInfo
@@ -336,8 +337,8 @@ abstract class IntelliJPlatformBasePlugin : Plugin<Project> {
                         dependencies,
                         downloadDirectory,
                         extensionProvider,
-                        gradle,
                         providers,
+                        project,
                         resources,
                     )
                 }
@@ -358,14 +359,17 @@ abstract class IntelliJPlatformBasePlugin : Plugin<Project> {
                 repositories,
                 dependencies,
                 providers,
-                gradle,
+                layout,
+                rootDir,
             )
 
             repositories.configureExtension<IntelliJPlatformRepositoriesExtension>(
                 Extensions.INTELLIJ_PLATFORM,
                 repositories,
                 providers,
-                gradle,
+                providers.localPlatformArtifactsDirectory(
+                    rootDir
+                ),
             )
         }
     }

--- a/src/main/kotlin/org/jetbrains/intellij/platform/gradle/plugins/settings/IntelliJPlatformSettingsPlugin.kt
+++ b/src/main/kotlin/org/jetbrains/intellij/platform/gradle/plugins/settings/IntelliJPlatformSettingsPlugin.kt
@@ -9,6 +9,7 @@ import org.jetbrains.intellij.platform.gradle.Constants.Extensions
 import org.jetbrains.intellij.platform.gradle.Constants.PLUGIN_SETTINGS_ID
 import org.jetbrains.intellij.platform.gradle.extensions.IntelliJPlatformRepositoriesExtension
 import org.jetbrains.intellij.platform.gradle.plugins.checkGradleVersion
+import org.jetbrains.intellij.platform.gradle.extensions.localPlatformArtifactsDirectory
 import org.jetbrains.intellij.platform.gradle.plugins.configureExtension
 import org.jetbrains.intellij.platform.gradle.utils.Logger
 import javax.inject.Inject
@@ -25,13 +26,16 @@ abstract class IntelliJPlatformSettingsPlugin @Inject constructor(
 
         checkGradleVersion()
 
+        val rootDir = settings.gradle.startParameter.projectDir ?: settings.rootDir
+        val localPlatformArtifactsDirectory = providers.localPlatformArtifactsDirectory(rootDir)
+
         @Suppress("UnstableApiUsage")
         with(settings.dependencyResolutionManagement.repositories) {
             configureExtension<IntelliJPlatformRepositoriesExtension>(
                 Extensions.INTELLIJ_PLATFORM,
                 this,
                 providers,
-                settings.gradle,
+                localPlatformArtifactsDirectory,
             )
         }
     }

